flink-end-to-end-perf-tests/flink-basic-operations/src/main/java/org/apache/flink/basic/utils/FromElements.java

```diff
@@ -0,0 +1,229 @@
+import org.apache.flink.api.java.typeutils.TypeExtractor;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.Meter;
+import org.apache.flink.metrics.MeterView;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.SimpleCounter;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+import org.apache.flink.streaming.api.datastream.DataStreamSource;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.source.ParallelSourceFunction;
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.flink.streaming.api.functions.source.SourceFunction;
+import org.apache.flink.util.Preconditions;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Provides the default data sets used for the WordCount example program. The default data sets are used, if no
+ * parameters are given to the program.
+ */
+public class FromElements {
+
+	@SafeVarargs
+	public static <OUT> DataStreamSource<OUT> fromElements(StreamExecutionEnvironment env, Long maxCount, int sleepNum, OUT... data) {
+		if (data.length == 0) {
+			throw new IllegalArgumentException("fromElements needs at least one element as argument");
+		}
+		TypeInformation<OUT> typeInfo;
+		try {
+			typeInfo = TypeExtractor.getForObject(data[0]);
+		} catch (Exception e) {
+			throw new RuntimeException("Could not create TypeInformation for type " + data[0].getClass().getName() + "; please specify the TypeInformation manually via StreamExecutionEnvironment#fromElements(Collection, TypeInformation)");
+		}
+		Collection<OUT> dataCollection = Arrays.asList(data);
+		FromElementsRichFunction.checkCollection(dataCollection, typeInfo.getTypeClass());
+		try {
+			SourceFunction<OUT> function =
+				new FromElementsRichFunction<>(typeInfo.createSerializer(env.getConfig()),
+					maxCount, sleepNum, dataCollection);
+			return env.addSource(function, "WordsData", typeInfo);
+		} catch (IOException e) {
+			throw new RuntimeException(e.getMessage(), e);
+		}
+	}
+
+	/**
+	 * A stream source function that returns a sequence of elements.
+	 *
+	 * <p>
+	 * Upon construction, this source function serializes the elements using Flink's type information. That way, any
+	 * object transport using Java serialization will not be affected by the serializability of the elements.
```

Comment:
Is this actually something we have to worry about? What use-case do we have for non-serializable data? Couldn't we simplify this class _a lot_ for mandating them to be serializable?

Reply:
I refactored and simplified this class.please interview it again,thanks